src/lowering/desugar.ml

```diff
@@ -457,6 +457,30 @@ and export_footprint self_id expr =
   )],
   [{ it = I.{ name = lab; var = v }; at = no_region; note = typ }])
 
+and export_gc_trigger self_id =
+  let open T in
+  let {lab;typ;_} = motoko_gc_trigger_fld in
+  let v = "$"^lab in
+  let scope_con1 = Cons.fresh "T1" (Abs ([], scope_bound)) in
+  let scope_con2 = Cons.fresh "T2" (Abs ([], Any)) in
+  let bind1  = typ_arg scope_con1 Scope scope_bound in
+  let bind2 = typ_arg scope_con2 Scope scope_bound in
+  let caller = fresh_var "caller" caller in
+  ([ letD (var v typ) (
+        funcE v (Shared Write) Promises [bind1] [] [] (
+            (asyncE T.Fut bind2
```

Comment:
Looking at the generated code, where the collector is called with a non-empty stack, with some Motoko values on the stack, I think it might be much safer to just do this in the backend only, in the fashion of export_async_helper. 

If that doesn't make sense, I can draft a follow on PR to clarify.

BTW, I also noticed that the primitive PerformGC doesn't update the mutator/collector stats correctly, leading to some  minor misaccounting. I think. Perhaps I missed that when adding the stats (just before you joined IIRC).

Reply:
Thanks a lot. The counting is indeed not correct. Indeed, it is much better to implement it in the backend. Many thanks also for the PR https://github.com/dfinity/motoko/pull/4303. I added some minor refinements on top of that PR https://github.com/dfinity/motoko/pull/4305.